File: src/lib/ladderSyncOps.ts
import {
  applyAttemptOutcome,
  type AttemptOutcome,
  type LadderState,
} from '@/lib/exposureLadder';

/**
 * US-609: the replay contract for offline exposure-ladder writes.
 *
 * A kitchen with bad signal is the normal case for this feature, so a logged
 * exposure has to survive the app being closed before it reaches Supabase.
 *
 * STATUS: this is the SPECIFICATION, not a wired-up code path. It has no
 * runtime caller yet, and the story's original plan -- route ladder writes
 * through app/mobile/lib/syncQueue -- turned out to target the wrong stack.
 * The Expo app under app/ has no ladder screen at all; the live iOS ladder is
 * native Swift (ios/EatPal/EatPal/Ladder/ExposureLadderPolicy.swift, writing
 * via Services/DataService.swift), and the offline queue it should use is
 * Services/OfflineStore.swift, whose Table enum today covers only foods, kids,
 * recipes, plan_entries and grocery_items. Ladder writes therefore cannot be
 * queued on iOS at all right now: an exposure logged offline is simply lost.
 *
 * What lives here is the part of that fix worth pinning down in one place and
 * testing before it is written twice: what a replay writes, given what the
 * server holds by the time it runs. exposureLadder.ts and
 * ExposureLadderPolicy.swift already mirror each other rung for rung, so the
 * Swift implementation should mirror these rules the same way.
 *
 * WHY THE QUEUE CARRIES INTENT, NOT THE COMPUTED ROW
 *
 * quickLog computes the next ladder state locally and PUTs it. Queueing that
 * computed row would mean replaying a snapshot built from state that may be
 * hours stale: if the other parent logged an attempt from their phone in the
 * meantime, the replay would stamp the older rung back over the newer one.
 * That is the "resurrecting a stale rung" the story rules out.
 *
 * So `ladder.attempt` carries the INTENT -- which rung was offered, what the
 * child did -- and the replay re-derives the next state from whatever the
 * server row says at replay time. applyAttemptOutcome is pure, so re-deriving
 * is exact: the local step re-applies on top of the server's state, which is
 * the conflict rule the story asks for and matches the load-precedence
 * contract in CLAUDE.md (a successful server read always wins; queued local
 * writes then re-apply).
 *
 * IDEMPOTENCY
 *
 * The attempt row carries a CLIENT-GENERATED uuid. food_attempts.id is
 * `UUID PRIMARY KEY DEFAULT gen_random_uuid()` and its Insert type takes an
 * optional id, so the client can pick it without a migration. The replay
 * checks for that id before doing anything: if the attempt is already there,
 * the whole op is treated as applied and nothing is written. That is what
 * stops a replayed attempt creating a duplicate food_attempts row or
 * advancing the ladder twice.
 *
 * KNOWN GAP, DELIBERATELY CHOSEN
 *
 * The insert and the ladder update are two statements, not one transaction.
 * If the process dies between them, a later replay sees the attempt already
 * present and skips the ladder move, so the rung stays where it was. The
 * exposure is still recorded and the next attempt recomputes from the real
 * server state, so the ladder self-corrects on the following log. This errs
 * toward under-applying rather than double-advancing, which is the direction
 * the story explicitly prioritises -- a ladder that skipped forward puts a
 * child in front of a food they have not earned yet, while one that lagged a
 * step only costs one extra exposure. Closing the gap entirely needs the two
 * writes in one RPC, which is a schema change and is not required here.
 */

/** The intent of one quick log, everything a replay needs to redo it. */
export interface LadderAttemptOp {
  /** Client-generated uuid; the dedupe key that makes replay exactly-once. */
  attemptId: string;
  /** kid_food_ladder.id -- the row to move. */
  ladderRowId: string;
  kidId: string;
  foodId: string;
  /** The rung the child was actually asked for, not the rung they end on. */
  stage: string;
  outcome: AttemptOutcome;
  /** ISO timestamp of the attempt itself, not of the replay. */
  attemptedAt: string;
  /** Local date the attempt belongs to; drives the next-due arithmetic. */
  today: string;
  mealSlot?: string | null;
  preparationMethod?: string | null;
  /** Plan entry to back-link, when the log came from the planner. */
  planEntryId?: string | null;
}

/** A direct parent edit: pause, resume, skip. Absolute, not derived. */
export interface LadderPatchOp {
  ladderRowId: string;
  patch: Partial<LadderState>;
}

/** What a replay decided to do. */
export type LadderReplayPlan =
  | { action: 'skip'; reason: 'already-applied' }
  | { action: 'skip'; reason: 'row-gone' }
  | { action: 'apply'; nextState: LadderState };

/**
 * Decide what a queued attempt should write, given what the server holds now.
 *
 * `serverState` is the CURRENT kid_food_ladder row (null when the row has since
 * been deleted); `attemptExists` is whether food_attempts already has this
 * op's attemptId. Kept free of Supabase so the conflict rule can be tested
 * directly.
 */
export function planAttemptReplay(args: {
  op: LadderAttemptOp;
  serverState: LadderState | null;
  attemptExists: boolean;
}): LadderReplayPlan {
  const { op, serverState, attemptExists } = args;

  // Already recorded: the op ran before, possibly in a previous session.
  // Re-deriving now would advance a ladder that already moved for it.
  if (attemptExists) {
    return { action: 'skip', reason: 'already-applied' };
  }

  // The food was removed from the ladder while the phone was offline. The
  // parent's later decision wins over the earlier queued exposure.
  if (!serverState) {
    return { action: 'skip', reason: 'row-gone' };
  }

  // Re-derive from the SERVER's state, never from the state captured offline.
  return {
    action: 'apply',
    nextState: applyAttemptOutcome(serverState, op.outcome, {
      today: op.today,
      attemptAt: op.attemptedAt,
    }),
  };
}

/** The food_attempts row for a queued attempt, including its dedupe id. */
export function attemptRowFromOp(op: LadderAttemptOp): Record<string, unknown> {
  return {
    id: op.attemptId,
    kid_id: op.kidId,
    food_id: op.foodId,
    stage: op.stage,
    outcome: op.outcome,
    attempted_at: op.attemptedAt,
    meal_slot: op.mealSlot ?? null,
    preparation_method: op.preparationMethod ?? null,
  };
}

/**
 * A queued op's payload is `Record<string, unknown>` by the time it comes back
 * out of storage, and storage is shared with older app builds. Validate before
 * replaying rather than trusting the shape.
 */
export function isLadderAttemptOp(value: unknown): value is LadderAttemptOp {
  if (!value || typeof value !== 'object') return false;
  const op = value as Record<string, unknown>;
  return (
    typeof op.attemptId === 'string' &&
    op.attemptId.length > 0 &&
    typeof op.ladderRowId === 'string' &&
    typeof op.kidId === 'string' &&
    typeof op.foodId === 'string' &&
    typeof op.outcome === 'string' &&
    typeof op.attemptedAt === 'string' &&
    typeof op.today === 'string'
  );
}

File: ios/EatPal/EatPal/Ladder/LadderSyncOps.swift
import Foundation

/// US-609: what a queued exposure replays, given what the server holds by the
/// time it runs.
///
/// A kitchen with bad signal is the normal case for this feature, so a logged
/// exposure has to survive the app being closed before it reaches Supabase.
/// Until now it did not: `OfflineStore.Table` covered foods, kids, recipes,
/// plan_entries and grocery_items, so a ladder write could not be queued at
/// all and an attempt logged offline was simply lost.
///
/// These rules are the Swift half of `src/lib/ladderSyncOps.ts`, which was
/// written and tested first precisely so they would not be derived twice --
/// the same arrangement `ExposureLadderPolicy` already has with
/// `src/lib/exposureLadder.ts`. Pure on purpose: the conflict rule is the part
/// worth testing, and it should not need a Supabase client to exercise.
///
/// WHY THE QUEUE CARRIES INTENT, NOT THE COMPUTED ROW
///
/// `quickLogExposure` computes the next ladder state locally and PUTs it.
/// Queueing that computed row would replay a snapshot built from state that
/// may be hours stale: if the other parent logged an attempt from their phone
/// in the meantime, the replay would stamp the older rung back over the newer
/// one. So the op carries the INTENT -- which rung was offered, what the child
/// did -- and the replay re-derives from whatever the server row says at
/// replay time. `ExposureLadderPolicy.apply` is pure, so re-deriving is exact.
/// That matches the load-precedence contract in CLAUDE.md: a successful server
/// read always wins, and queued local writes then re-apply on top of it.
///
/// KNOWN GAP, DELIBERATELY CHOSEN
///
/// The attempt insert and the ladder update are two statements, not one
/// transaction. If the process dies between them, a later replay sees the
/// attempt already present and skips the ladder move, so the rung stays where
/// it was. The exposure is still recorded and the next log recomputes from
/// real server state, so the ladder self-corrects. This errs toward
/// under-applying rather than double-advancing, which is the direction that
/// matters here: a ladder that skipped forward puts a child in front of a food
/// they have not earned, while one that lagged costs one extra exposure.
/// Closing it entirely needs both writes in one RPC.

/// The intent of one quick log: everything a replay needs to redo it.
///
/// `attemptId` is generated on the client and is the dedupe key that makes
/// replay exactly-once. `food_attempts.id` is `UUID PRIMARY KEY DEFAULT
/// gen_random_uuid()` and accepts a supplied value, so this needs no
/// migration -- proved against PostgreSQL in
/// `supabase/diagnostics/us-609-attempt-replay-check.sql`.
struct LadderAttemptOp: Codable, Equatable {
    /// Client-generated uuid; the dedupe key.
    var attemptId: String
    /// `kid_food_ladder.id` -- the row to move.
    var ladderRowId: String
    var kidId: String
    var foodId: String
    /// The rung the child was actually asked for, not the rung they end on.
    var stage: String
    var outcome: LadderOutcome
    /// ISO timestamp of the attempt itself, not of the replay.
    var attemptedAt: String
    /// Local date the attempt belongs to; drives the next-due arithmetic.
    /// Captured at log time so a replay days later does not schedule from the
    /// wrong day.
    var today: String
    var mealSlot: String?
    var preparationMethod: String?
    /// Plan entry to back-link, when the log came from the planner.
    var planEntryId: String?
}

/// What a replay decided to do.
enum LadderReplayPlan: Equatable {
    /// The attempt is already on the server: it ran before, possibly in a
    /// previous session. Re-deriving would advance a ladder that already moved.
    case skipAlreadyApplied
    /// The food was removed from the ladder while the phone was offline. The
    /// parent's later decision wins over the earlier queued exposure.
    case skipRowGone
    case apply(LadderState)
}

enum LadderSyncOps {
    /// Decide what a queued attempt should write, given what the server holds
    /// now.
    ///
    /// `serverState` is the CURRENT `kid_food_ladder` row, nil when it has
    /// since been deleted. `attemptExists` is whether `food_attempts` already
    /// carries this op's `attemptId`.
    static func planAttemptReplay(
        op: LadderAttemptOp,
        serverState: LadderState?,
        attemptExists: Bool
    ) -> LadderReplayPlan {
        if attemptExists { return .skipAlreadyApplied }
        guard let serverState else { return .skipRowGone }

        // Re-derive from the SERVER's state, never from the state captured
        // offline. This is the whole conflict rule.
        return .apply(
            ExposureLadderPolicy.apply(
                serverState,
                outcome: op.outcome,
                today: op.today,
                attemptAt: op.attemptedAt
            )
        )
    }

    /// The `food_attempts` row for a queued attempt, carrying its dedupe id.
    static func attemptRow(from op: LadderAttemptOp) -> FoodAttemptInsert {
        FoodAttemptInsert(
            id: op.attemptId,
            kidId: op.kidId,
            foodId: op.foodId,
            stage: op.stage,
            outcome: op.outcome.rawValue,
            attemptedAt: op.attemptedAt,
            mealSlot: op.mealSlot,
            preparationMethod: op.preparationMethod
        )
    }

    /// Whether a failed write hit a uniqueness constraint.
    ///
    /// Both of this feature's idempotency signals are unique violations, and
    /// they are DIFFERENT constraints on different keys: `food_attempts.id`
    /// (primary key) and `kid_food_ladder (kid_id, food_id)` (unique index).
    /// Both must be swallowed as success. A queue that retries one spins on an
    /// already-applied op forever, which is worse than the duplicate it is
    /// trying to avoid.
    ///
    /// Matched on the rendered error rather than a typed `PostgrestError`
    /// field: the concrete error type supabase-swift surfaces has changed
    /// shape across 2.x, and this must not start silently returning false on a
    /// dependency bump. Both the SQLSTATE and Postgres's own wording are
    /// checked so either one alone is enough.
    static func isUniqueViolation(_ error: Error) -> Bool {
        isUniqueViolation(describing: String(describing: error))
    }

    /// Split out so the matching can be tested without constructing a
    /// PostgREST error, which needs the network layer.
    static func isUniqueViolation(describing text: String) -> Bool {
        let lowered = text.lowercased()
        return lowered.contains("23505") || lowered.contains("duplicate key")
    }
}
